magefiles/tests.go

```diff
@@ -45,25 +44,50 @@ func (Test) Python() error {
 		return nil
 	}
 
+	args := []string{
+		"--confcutdir=.",
+	}
+	args = append(args, testFiles...)
+
+	// Add testName filter if provided
+	if testName != "" {
+		args = append(args, "-k", testName, "-vv")
+	} else {
+		args = append(args, "-k", "not [file")
+	}
+
 	//  Run the tests (currently just the server ones)
 	if err := sh.RunWithV(map[string]string{
 		"MLFLOW_GO_LIBRARY_PATH": libpath,
-	}, "pytest",
-		"--confcutdir=.",
-		".mlflow.repo/tests/tracking/test_rest_tracking.py",
-		".mlflow.repo/tests/tracking/test_model_registry.py",
-		".mlflow.repo/tests/store/tracking/test_sqlalchemy_store.py",
-		".mlflow.repo/tests/store/model_registry/test_sqlalchemy_store.py",
-		"-k",
-		"not [file",
-		// "-vv",
+	}, "pytest", args...,
+	// "-vv",
 	); err != nil {
 		return err
 	}
 
 	return nil
 }
 
+// Run mlflow Python tests against the Go backend.
+func (Test) Python() error {
+	return runPythonTests([]string{
+		".mlflow.repo/tests/tracking/test_rest_tracking.py",
+		".mlflow.repo/tests/tracking/test_model_registry.py",
+		".mlflow.repo/tests/store/tracking/test_sqlalchemy_store.py",
+		".mlflow.repo/tests/store/model_registry/test_sqlalchemy_store.py",
+	}, "")
+}
+
+// Run specific Python test against the Go backend.
+func (Test) PythonSpecific(testName string) error {
```

Comment:
I added all files so that it wouldn't miss some of the tests in other files (at first I was not testing comprehensively because I was only looking at the `test_sqlalchemy_store.py`). It matches multiple files so that you can test anything related to your endpoint in one go.

Reply:
@dsuhinin a mage target cannot have optional arguments.
Maybe we should change the target slightly to take a single string argument and we can pass all pytest arguments in there. When you want to run a specific test you can then do `mage pythonSpecific "filename.py::test-name -vv"`

Reply:
I have been working on implementing the single string argument testing without much luck. I'm currently working in a separate branch as I was hoping this could get merged in even if the test scripts are not perfect! :)

This I have right now in `tests.go` if you'd like to test it. I'm having issues with the filename matching, right now it keeps executing all the tests instead of just the ones I ask for:
```go
//go:build mage

//nolint:wrapcheck
package main

import (
	"os"

	"github.com/magefile/mage/mg"
	"github.com/magefile/mage/sh"
)

type Test mg.Namespace

func cleanUpMemoryFile() error {
	// Clean up :memory: file
	filename := ":memory:"
	_, err := os.Stat(filename)

	if err == nil {
		// File exists, delete it
		err = os.Remove(filename)
		if err != nil {
			return err
		}
	}

	return nil
}

func runPythonTests(testFiles []string, testName string) error {
	libpath, err := os.MkdirTemp("", "")
	if err != nil {
		return err
	}

	// Remove the Go binary
	defer os.RemoveAll(libpath)
	//nolint:errcheck
	defer cleanUpMemoryFile()

	// Build the Go binary in a temporary directory
	if err := sh.RunV("python", "-m", "mlflow_go.lib", ".", libpath); err != nil {
		return nil
	}

	args := []string{
		"--confcutdir=.",
	}

	// Use testFiles directly if specified, otherwise use defaults
	if len(testFiles) > 0 {
		args = append(args, testFiles...)
	} else {
		args = append(args, ".mlflow.repo/tests/tracking/test_rest_tracking.py",
			".mlflow.repo/tests/tracking/test_model_registry.py",
			".mlflow.repo/tests/store/tracking/test_sqlalchemy_store.py",
			".mlflow.repo/tests/store/model_registry/test_sqlalchemy_store.py")
	}

	// If testName is provided, append it directly
	if testName != "" {
		args = append(args, testName) // No need for -k if full test path is used
	}

	//  Run the tests (currently just the server ones)
	if err := sh.RunWithV(map[string]string{
		"MLFLOW_GO_LIBRARY_PATH": libpath,
	}, "pytest", args...); err != nil {
		return err
	}

	return nil
}

// Run mlflow Python tests against the Go backend.
func (Test) Python(pytestArgs string) error {
	return runPythonTests([]string{
		".mlflow.repo/tests/tracking/test_rest_tracking.py",
		".mlflow.repo/tests/tracking/test_model_registry.py",
		".mlflow.repo/tests/store/tracking/test_sqlalchemy_store.py",
		".mlflow.repo/tests/store/model_registry/test_sqlalchemy_store.py",
	}, pytestArgs)
}

// Run specific Python test against the Go backend.
func (Test) PythonSpecific(pytestArgs string) error {
	return runPythonTests([]string{
		".mlflow.repo/tests/tracking/test_rest_tracking.py",
		".mlflow.repo/tests/tracking/test_model_registry.py",
		".mlflow.repo/tests/store/tracking/test_sqlalchemy_store.py",
		".mlflow.repo/tests/store/model_registry/test_sqlalchemy_store.py",
	}, pytestArgs)
}

// Run the Go unit tests.
func (Test) Unit() error {
	return sh.RunV("go", "test", "./pkg/...")
}

// Run all tests.
func (Test) All() {
	mg.Deps(Test.Unit, func() error {
		return Test{}.Python("")
	})
}
```

Reply:
I refactored this a little, tl;dr you can pass all the files I think.
It makes more sense to accept an argument like `pytest` would expect.

Reply:
Ohh, I got it now, thanks @nojaf! Feel free to merge the PR now as I don't have write access.